homeassistant/components/homematicip_cloud.py

```diff
@@ -5,137 +5,186 @@
+        vol.Optional(CONF_NAME, default=None): vol.Any(None, cv.string),
         vol.Required(CONF_ACCESSPOINT): cv.string,
         vol.Required(CONF_AUTHTOKEN): cv.string,
     })],
 }, extra=vol.ALLOW_EXTRA)
 
-EVENT_HOME_CHANGED = 'homematicip_home_changed'
-EVENT_DEVICE_CHANGED = 'homematicip_device_changed'
-EVENT_GROUP_CHANGED = 'homematicip_group_changed'
-EVENT_SECURITY_CHANGED = 'homematicip_security_changed'
-EVENT_JOURNAL_CHANGED = 'homematicip_journal_changed'
+HMIP_ACCESS_POINT = 'Access Point'
+HMIP_HUB = 'HmIP-HUB'
 
 ATTR_HOME_ID = 'home_id'
-ATTR_HOME_LABEL = 'home_label'
+ATTR_HOME_NAME = 'home_name'
 ATTR_DEVICE_ID = 'device_id'
 ATTR_DEVICE_LABEL = 'device_label'
 ATTR_STATUS_UPDATE = 'status_update'
 ATTR_FIRMWARE_STATE = 'firmware_state'
+ATTR_UNREACHABLE = 'unreachable'
 ATTR_LOW_BATTERY = 'low_battery'
+ATTR_MODEL_TYPE = 'model_type'
+ATTR_GROUP_TYPE = 'group_type'
+ATTR_DEVICE_RSSI = 'device_rssi'
+ATTR_DUTY_CYCLE = 'duty_cycle'
+ATTR_CONNECTED = 'connected'
 ATTR_SABOTAGE = 'sabotage'
-ATTR_RSSI = 'rssi'
-ATTR_TYPE = 'type'
+ATTR_OPERATION_LOCK = 'operation_lock'
+
+
+class HomematicipConnector:
+    """Manages HomematicIP http and websocket connection."""
+
+    def __init__(self, config, websession, hass):
+        """Initialize HomematicIP cloud connection."""
+        from homematicip.async.home import AsyncHome
+        self._hass = hass
+        self._ws_close_requested = False
+        self._retry_task = None
+        self._tries = 0
+        self._accesspoint = config.get(CONF_ACCESSPOINT)
+        self._authtoken = config.get(CONF_AUTHTOKEN)
 
+        self.home = AsyncHome(hass.loop, websession)
+        self.home.set_auth_token(self._authtoken)
 
-def setup(hass, config):
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.close())
+
+    async def init(self):
+        """Initial connection."""
+        await self.home.init(self._accesspoint)
+        await self.home.get_current_state()
+
+    async def _handle_connection(self):
+        """Handle websocket connection."""
+        from homematicip.base.base_connection import HmipConnectionError
+
+        await self.home.get_current_state()
+        hmip_events = await self.home.enable_events()
+        try:
+            await hmip_events
+        except HmipConnectionError:
+            return
+        return
+
+    async def connect(self):
+        """Start websocket connection."""
+        self._tries = 0
+        while True:
+            try:
+                await self._handle_connection()
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected error")
+            if self._ws_close_requested:
+                break
+            self._ws_close_requested = False
+            self._tries += 1
+            try:
+                self._retry_task = self._hass.async_add_job(asyncio.sleep(
+                    2 ** min(9, self._tries), loop=self._hass.loop))
```

Comment:
You need to wait the _retry_task otherwise the sleep command just wanders off and does not make the loop wait.

```await self._retry_task```

Reply:
Thanks!